File: 3 - AG/Cromossomo.java

class Cromossomo implements Comparable<Cromossomo> {

    StringBuffer valor;
    int aptidao;
    int porcentagemAptidao;

    public Cromossomo(StringBuffer valor, String estadoFinal) {
        this.valor = valor;
        this.aptidao = calcularAptidao(estadoFinal);
    }

    //heuristica
    int calcularAptidao(String estadoFinal) {
        int nota = 0;
        for (int i = 0; i < estadoFinal.length(); i++) {
            if (this.valor.toString().contains(estadoFinal.charAt(i) + "")) {
                nota += 5;
            }
            if (this.valor.toString().charAt(i) == estadoFinal.charAt(i)) {
                nota += 50;
            }
        }
        return nota;
    }

    public int compareTo(Cromossomo cromossomo) {
        if (this.aptidao > cromossomo.aptidao) {
            return -1;
        }
        return 1;
    }
}

File: 3 - AG/python/cromossomo.py
class Cromossomo:

    def __init__(self, valor, estado_final):
        self.valor = valor
        self.aptidao = self.calcular_aptidao(estado_final)
    
    def calcular_aptidao(self, estado_final):
        nota = 0
        for i in range(len(estado_final)):
            if (estado_final[i] in self.valor):
               nota += 5
               
            if (self.valor[i] == estado_final[i]):
                nota += 50
            
        return nota
    
    def __eq__(self, other):
        if isinstance(other, Cromossomo):
            return self.valor == other.valor
        return False
    
    def __gt__(self, other):
        return self.aptidao <= other.aptidao
    
    def __str__(self):
        return f"valor= {self.valor}, aptidao= {self.aptidao}"
        # return "valor=" + str(self.valor) + ", aptidao=" + str(self.aptidao )